golang/cosmos/x/swingset/keeper/extension_snapshotter.go

```diff
@@ -0,0 +1,270 @@
+	newRestoreContext       func(height int64) sdk.Context
+	swingStoreSnapshotter   *SwingStoreSnapshotter
+	getSwingStoreExportData func(ctx sdk.Context) []*vstoragetypes.DataEntry
+	logger                  log.Logger
+	activeExtensionSnapshot *activeExtensionSnapshot
+}
+
+// NewSwingsetSnapshotter creates a new SwingsetSnapshotter
+func NewSwingsetSnapshotter(app *baseapp.BaseApp, swingStoreSnapshotter *SwingStoreSnapshotter, getSwingStoreExportData func(ctx sdk.Context) []*vstoragetypes.DataEntry) SwingsetSnapshotter {
+	return SwingsetSnapshotter{
+		isConfigured:    func() bool { return app.SnapshotManager() != nil },
+		takeAppSnapshot: app.Snapshot,
+		newRestoreContext: func(height int64) sdk.Context {
+			return app.NewUncachedContext(false, tmproto.Header{Height: height})
+		},
+		logger:                  app.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName), "submodule", "extension snapshotter"),
+		swingStoreSnapshotter:   swingStoreSnapshotter,
+		getSwingStoreExportData: getSwingStoreExportData,
+		activeExtensionSnapshot: nil,
+	}
+}
+
+// SnapshotName returns the name of snapshotter, it should be unique in the manager.
+// Implements ExtensionSnapshotter
+func (snapshotter *SwingsetSnapshotter) SnapshotName() string {
+	return types.ModuleName
+}
+
+// SnapshotFormat returns the default format the extension snapshotter uses to encode the
+// payloads when taking a snapshot.
+// It's defined within the extension, different from the global format for the whole state-sync snapshot.
+// Implements ExtensionSnapshotter
+func (snapshotter *SwingsetSnapshotter) SnapshotFormat() uint32 {
+	return SnapshotFormat
+}
+
+// SupportedFormats returns a list of formats it can restore from.
+// Implements ExtensionSnapshotter
+func (snapshotter *SwingsetSnapshotter) SupportedFormats() []uint32 {
+	return []uint32{SnapshotFormat}
+}
+
+// WaitUntilSnapshotStarted synchronizes with a snapshot in progress, if any.
+// The app must call this method before sending a commit action to SwingSet after
+// initiating a snapshot.
+func (snapshotter *SwingsetSnapshotter) WaitUntilSnapshotStarted() error {
+	return snapshotter.swingStoreSnapshotter.WaitUntilSnapshotStarted()
+}
+
+// InitiateSnapshot initiates a snapshot for the given height.
+// If a snapshot is already in progress, or if no snapshot manager is
+// configured, this will fail.
+// The snapshot operation is performed in a goroutine managed by the shared
+// SwingStoreSnapshotter. Use `WaitUntilSnapshotStarted` to synchronize commit
+// boundaries.
+func (snapshotter *SwingsetSnapshotter) InitiateSnapshot(height int64) error {
+	if !snapshotter.isConfigured() {
+		return fmt.Errorf("snapshot manager not configured")
+	}
+
+	return snapshotter.swingStoreSnapshotter.InitiateSnapshot(uint64(height), snapshotter, SwingStoreExportOptions{
+		ExportMode:        "current",
+		IncludeExportData: false,
+	})
+}
+
+// TakeSnapshot performs the app snapshot, which will call `SnapshotExtension`
```

Comment:
Can the difference between "initiating" vs. "performing"/"taking" a snapshot be better documented? I wonder if the latter is even the right verb as opposed to something like "finish" or "populate".